solidity/nft_token.sol

```diff
@@ -59,190 +60,245 @@ contract nft_token {
+        // Transfer bridged tokens to the user
         SplToken.mint_to_pda(
-            mintErc20,
-            tx.accounts.toErc20.key,
+            bridgedTokenMint,
+            tx.accounts.userBridgedTokenAccount.key,
             pdaProgram,
-            liquidity,
+            positionLiquidity,
             pdaProgramSeed,
             pdaBump);
 
-        // Record liquidity and its correspondent token account address
-        address tokenAccount = tx.accounts.pdaTokenAccount.key;
-        mapTokenAccountLiquidity[tokenAccount] = liquidity;
-        tokenAccounts[numTokenAccounts] = (tx.accounts.pdaTokenAccount.key);
-        numTokenAccounts++;
+        // Record position liquidity amount and its correspondent account address
+        address positionAddress = tx.accounts.position.key;
+        mapPositionAccountLiquidity[positionAddress] = positionLiquidity;
+        address pdaPositionAta = tx.accounts.pdaPositionAccount.key;
+        mapPositionAccountPdaAta[positionAddress] = pdaPositionAta;
+        positionAccounts[numPositionAccounts] = positionAddress;
+        numPositionAccounts++;
     }
 
-    @mutableAccount(fromERC20Account)
-    @mutableAccount(pdaERC20Account)
-    @mutableAccount(fromWallet)
-    @mutableAccount(pdaTokenAccount)
-    @mutableAccount(fromTokenAccount)
-    @mutableAccount(mintERC20)
+    @mutableAccount(pool)
+    @mutableAccount(tokenProgramId)
+    @mutableAccount(position)
+    @mutableAccount(userBridgedTokenAccount)
+    @mutableAccount(pdaBridgedTokenAccount)
+    @mutableAccount(userWallet)
+    @mutableAccount(userPositionAccount)
+    @mutableAccount(bridgedTokenMint)
+    @mutableAccount(pdaPositionAccount)
+    @mutableAccount(userTokenAccountA)
+    @mutableAccount(userTokenAccountB)
+    @mutableAccount(tokenVaultA)
+    @mutableAccount(tokenVaultB)
+    @mutableAccount(tickArrayLower)
+    @mutableAccount(tickArrayUpper)
+    @mutableAccount(positionMint)
     @signer(sig)
     // Transfer with PDA
     function withdraw(uint64 amount) external {
+        address positionAddress = positionAccounts[firstAvailablePositionAccountIndex];
+        if (positionAddress != tx.accounts.position.key) {
+            revert("Wrong liquidity token account");
+        }
+
+        address pdaPositionAta = tx.accounts.pdaPositionAccount.key;
+        if (mapPositionAccountPdaAta[positionAddress] != pdaPositionAta) {
+            revert("Wrong position ATA");
+        }
+
+        uint64 positionLiquidity = mapPositionAccountLiquidity[positionAddress];
         // Check that the token account exists
-        if (mapTokenAccountLiquidity[tx.accounts.pdaTokenAccount.key] == 0) {
+        if (positionLiquidity == 0) {
             revert("No liquidity on a provided token account");
         }
 
-        // Transfer ERC20 tokens to the pdaERC20Account address of this program
+        // Check the requested amount to be smaller or equal than the position liquidity
+        if (amount > positionLiquidity) {
```

Comment:
I had in mind that in this case we considered the second-to-last nft, and so on. It is likely the the protocol will have an amount larger than any liquidity position

Reply:
We can only go one NFT at a time.